# Phase L — The gate, and the tests that matter either way

Phase K answered thirteen objections. A third referee report then showed that
one of those answers was structurally incapable of answering its question, and
that the question it *should* have answered decides whether H3 stays in the
paper at all.

## L1 is a gate, not a test

> **Is the 3.29 mm seasonal amplitude a property of the mat, or of the
> reference zone we happened to measure it against?**

Phase K drew its eight controls from zone D. The zone definitions make that
useless for this question:

```
D = outside & ~water & flat & ~C
```

D is the **complement** of the matched reference, so those controls were
guaranteed *not* land-cover matched. Their 2.5–10.2 mm spread measured how
much land cover varies — not how much our amplitude depends on the choice of a
**comparable** control. L1 uses `matched_cover_pool`, which rebuilds the pool
as same-class terrain with the published reference excluded.

**Run L1 first and read its verdict before running anything else.**

| If | Then |
|---|---|
| matched controls are tight and A − control clears the matched null | H3 survives, heavily caveated; L4–L8 are worth doing |
| matched controls are spread and A − control sits inside the null | **H3 falls.** The 3.29 mm is a control-selection artefact, comes out of the paper, and half of what follows becomes moot |

The second outcome is a real possibility, not a formality. In Phase K the
observed value sat at the 43rd percentile of unmatched controls — four of seven
exceeded it.

## The rest

| # | Test | Depends on the gate? | Cost |
|---|---|---|---|
| **L1** | **THE GATE** — matched controls + matched null | — | **[SLOW]** |
| L2 | Erode zone A: deficit, amplitude, closure on the pure interior | **no** | **[SLOW]** |
| L3 | Subdivide the reference at fixed target | **no** | **[SLOW]** |
| L4 | Closed-triplet count, two ways | **no** | instant |
| L5 | Seasonal amplitude fitted in the wrapped domain | if gate passes | fast |
| L6 | Corrected null raised to 5 000 draws | if gate passes | **[VERY SLOW]** |
| L7 | The 27 pixels, by autocorrelation-preserving permutation | **no** | fast |
| L8 | Analytic + residual bootstrap intervals | if gate passes | fast |

L2, L3, L4 and L7 support the **robust** results — the coherence deficit, the
closure dispersion, the per-pixel failure — which are the paper whichever way
the gate falls. Run them regardless.

## 0. Bootstrap

In [ ]:
# --- Repo bootstrap (the one cell that cannot use the package) -----------
from pathlib import Path
from google.colab import userdata, drive
BRANCH = "main"
repo_dir = Path("/content/SAr-adapted")
try:
    token = userdata.get("GITHUB_TOKEN")
except Exception:
    token = None
if repo_dir.exists():
    %cd {repo_dir}
    !git checkout {BRANCH} && git pull origin {BRANCH}
else:
    clone_url = f"https://x-access-token:{token}@github.com/AimenSayoud/SAr-adapted.git" if token else "https://github.com/AimenSayoud/SAr-adapted.git"
    !git clone --branch {BRANCH} {clone_url} {repo_dir}
    %cd {repo_dir}
!bash environment/colab_setup.sh
# Purge stale modules: git pull updates files, not modules already in memory.
import sys, importlib
sys.path.insert(0, str(repo_dir / "src"))
for _m in [m for m in list(sys.modules) if m.startswith("insar_wetlands")]:
    del sys.modules[_m]
importlib.invalidate_caches()


/content/SAr-adapted
Already on 'main'
Your branch is up to date with 'origin/main'.


From https://github.com/AimenSayoud/SAr-adapted
 * branch            main       -> FETCH_HEAD
Already up to date.


>>> Pas de lock file — installation des dependances pip (non epinglees)...


>>> Installation du package local insar_wetlands...


  Installing build dependencies ... done
  Checking if build backend supports build_editable ... done


  Getting requirements to build editable ... done
  Preparing editable metadata (pyproject.toml) ... done


  Building editable for insar-wetlands (pyproject.toml) ... done


>>> Environnement pret.


## 1. Data and zones

In [ ]:
# --- Phase context -------------------------------------------------------
# Replaces the ~30 lines of setup this notebook used to carry. Drive mount,
# config, path resolution, logging and the run archive all come from here, so a
# change to any of them is a one-file edit instead of 38.
import numpy as np, pandas as pd, xarray as xr, matplotlib.pyplot as plt
from insar_wetlands.bootstrap import start

ctx = start("phaseL")

# Familiar names, so the rest of the notebook is unchanged. Everything is lazy:
# `zones` triggers the water mask, WorldCover and the S2 features on first use.
cfg      = ctx.cfg
DRIVE    = ctx.paths.drive
CROPPED  = ctx.paths.cropped
OUTDIR   = ctx.outdir
to_grid  = ctx.to_grid
pairs    = ctx.pairs
template = ctx.template
dem      = ctx.dem
zones    = ctx.zones


INFO    phase phaseL starting


INFO      repo             /content/SAr-adapted


INFO      drive            /content/drive/MyDrive/insar_rzecin


INFO      phase            phaseL


INFO      on_colab         True


INFO      drive_exists     True


INFO      cropped_exists   True


WARNING git setup skipped: Secret 'GIT_USER_EMAIL' introuvable. L'ajouter dans Colab (icone cle) et autoriser l'acces pour ce notebook.


INFO    pairs: 356


INFO    template grid: {'y': 129, 'x': 138}


INFO    zones: {'A': 499, 'B': 65, 'C': 398, 'D': 10750}


In [ ]:
import matplotlib as mpl
import insar_wetlands.referee as ref
from insar_wetlands.zone_viz import save_figure, set_zone_labels, ZONE_COLORS
# --- retained from the original setup cell ---
from insar_wetlands.stack import (
    load_layer, load_static_layer, list_pairs,
    align_grid, dates_from_pairs, to_grid)
from insar_wetlands.stratify import (
    define_zones, load_worldcover,
    s2_landcover_features, signed_distance_to_aoi,
    coherence_by_zone_stream, paired_zone_diff)
from insar_wetlands.aggregate import (
    seasonal_amplitude, empirical_pvalue,
    null_distribution, closure_bias_by_zone,
    aggregate_unwrapped, invert_aggregate)
from insar_wetlands.zone_viz import set_zone_labels
from insar_wetlands.referee import verdict, VERDICTS
from insar_wetlands.masking.water_mask import flooded_fraction
from insar_wetlands.inversion.phaselinking import wrap_unw
set_zone_labels("en")
mpl.rcParams.update({"figure.dpi": 110, "savefig.dpi": 300, "font.size": 9,
                     "axes.grid": True, "grid.alpha": .3})
OUT = Path("docs/paper/referee"); OUT.mkdir(parents=True, exist_ok=True)
ff = to_grid(flooded_fraction(xr.open_dataset(DRIVE / "water_mask.nc")), template)
try: wc = load_worldcover(template, cfg)
except Exception as e: print("WorldCover:", e); wc = None
try: s2 = xr.load_dataset(DRIVE / "s2_stack.nc"); s2f = s2_landcover_features(s2)
except Exception as e: print("S2:", e); s2f = None
sd = signed_distance_to_aoi(template, cfg)
nA = int(zones["A"].values.sum()); nC = int(zones["C"].values.sum())
unw = load_layer(CROPPED, "unw_phase", pairs); corr = load_layer(CROPPED, "corr", pairs)
wrapped = wrap_unw(unw)          # closure statistics need wrapped phase


---

# L1 — THE GATE  *(R1 M5, corrected)*

Two distributions from **one** pool of land-cover-matched terrain:

- **A − control**: the mat measured against each of many matched controls.
  If the mat carries a real seasonal signal, every matched control should see
  it, so this should be tight and clearly displaced from zero.
- **control − control**: two matched patches against each other. This is the
  correct null — it fixes the land-cover mismatch *and* the geometry
  asymmetry R2 §4.5 identified, in one construction.

The comparison between them is the answer. Anything else is commentary.

In [ ]:
# The pool: same WorldCover class as the mat, outside the site, flat, unflooded,
# with the published reference excluded so the controls are independent of it.
assert wc is not None, 'L1 needs WorldCover; without it the pool cannot be built'
pool = ref.matched_cover_pool(zones, wc, exclude_reference='C')
print(f"matched pool: {pool.attrs['n_px']} px, WorldCover class "
      f"{pool.attrs['dominant_class']} | zone C = {nC} px, zone D = "
      f"{int(zones['D'].values.sum())} px")
need = nA + nC
print(f'null needs {need} px total (two disjoint patches); '
      f'pool has {pool.attrs["n_px"]}')
if pool.attrs['n_px'] < need:
    print('!! pool below the un-scaled requirement -- the null will be built',
          'from proportionally smaller patches, which makes it WIDER and the',
          'gate more conservative. That is reported, not hidden.')
if pool.attrs['n_px'] < nC:
    raise RuntimeError(f'matched pool has only {pool.attrs["n_px"]} px, '
                       f'below one control ({nC}). L1 cannot run; widen the '
                       'pool (e.g. exclude_reference=None) or relax the class '
                       'match, and say which in the paper.')
zp = {**zones, 'POOL': pool}

matched pool: 891 px, WorldCover class 30 | zone C = 398 px, zone D = 10750 px
null needs 897 px total (two disjoint patches); pool has 891
!! pool below the un-scaled requirement -- the null will be built from proportionally smaller patches, which makes it WIDER and the gate more conservative. That is reported, not hidden.


In [ ]:
# observed, and the two distributions
obs = seasonal_amplitude(invert_aggregate(
        aggregate_unwrapped(unw, corr, zones, 'A', 'C')))
print(f"observed A - C : {obs['amplitude_mm']:.3f} mm, phase DOY {obs['phase_doy']:.0f}")

mc = ref.multi_control_patches(unw, corr, zp, template, n_px=nC,
                               target='A', pool='POOL', n_controls=25)
# Two INDEPENDENT disjoint patches of the matched pool -- not two adjacent
# halves of one contiguous blob, which is what null_distribution builds and
# which needs nA+nC pixels of the mat's own land-cover class side by side.
# That demand is usually infeasible on a matched pool, and when it fails the
# frame comes back EMPTY and every statistic below raises with no clue why.
nullm = ref.matched_null_pairs(unw, corr, zp, template, 'POOL', nA, nC,
                               n_trials=300)
if nullm.attrs.get('size_scale', 1.0) != 1.0:
    print('!!', nullm.attrs['note'])
    print(f"   null patches: {nullm.attrs['n_target']} / "
          f"{nullm.attrs['n_reference']} px against the real {nA} / {nC}")
print(f'A - matched control : {len(mc)} controls')
print(f'matched null        : {len(nullm)} draws')
mc.to_csv(OUT/'LT01_matched_controls.csv', index=False)
nullm.to_csv(OUT/'LT02_matched_null.csv', index=False)

observed A - C : 3.286 mm, phase DOY 104


!! patch sizes scaled by 0.99 to fit the pool; the null is therefore conservative (wider)
   null patches: 495 / 395 px against the real 499 / 398
A - matched control : 25 controls
matched null        : 300 draws


In [ ]:
sig = mc.amplitude_mm.dropna().values; nul = nullm.amplitude_mm.dropna().values
if len(sig) < 5 or len(nul) < 20:
    raise RuntimeError(
        f'L1 has too little to compare: {len(sig)} controls, {len(nul)} null '
        f'draws. An empty null usually means the matched pool could not '
        f'supply the patch sizes -- check the pool report printed above.')
p95 = float(np.percentile(nul, 95)); nmed = float(np.median(nul))
clears = float((sig > p95).mean())
pv = empirical_pvalue(float(np.median(sig)), nul)

summ = pd.DataFrame([
  {'distribution':'A - matched control','n':len(sig),'median':np.median(sig),
   'iqr':np.percentile(sig,75)-np.percentile(sig,25),'p95':np.percentile(sig,95)},
  {'distribution':'matched null (ctrl-ctrl)','n':len(nul),'median':nmed,
   'iqr':np.percentile(nul,75)-np.percentile(nul,25),'p95':p95}])
display(summ.round(3)); summ.to_csv(OUT/'LT03_gate_summary.csv', index=False)

fig,ax=plt.subplots(figsize=(7,4))
ax.hist(nul,bins=40,alpha=.6,color='gray',density=True,label=f'matched null (n={len(nul)})')
ax.hist(sig,bins=20,alpha=.7,color=ZONE_COLORS['A'],density=True,
        label=f'A - matched control (n={len(sig)})')
ax.axvline(p95,ls='--',c='k',lw=1.3,label=f'null p95 = {p95:.2f} mm')
ax.axvline(obs['amplitude_mm'],c='tab:red',lw=2,label=f"published A-C = {obs['amplitude_mm']:.2f} mm")
ax.set_xlabel('seasonal amplitude (mm)'); ax.legend(fontsize=8)
ax.set_title('L1 — is the amplitude a property of the mat or of the reference?')
save_figure(fig,'L01_gate',OUT)

PASSED = (np.median(sig) > p95) and (clears >= 0.5)
print('='*74)
print(f"  A - matched control : median {np.median(sig):.2f} mm  "
      f"IQR {np.percentile(sig,75)-np.percentile(sig,25):.2f}")
print(f"  matched null        : median {nmed:.2f} mm  p95 {p95:.2f}")
print(f"  controls clearing the null p95 : {100*clears:.0f} %")
print(f"  empirical p of the median      : {pv['p_value']:.4f}")
print('='*74)
print(f"  GATE: {'PASSED' if PASSED else 'FAILED'}")
print('='*74)

verdict('L1', (
  f"Against {len(sig)} land-cover-matched controls the mat gives "
  f"{np.median(sig):.2f} mm (IQR {np.percentile(sig,75)-np.percentile(sig,25):.2f}); "
  f"the matched null has median {nmed:.2f} mm and p95 {p95:.2f} mm. "
  f"{100*clears:.0f} % of controls clear the null, empirical p = {pv['p_value']:.4f}. "
  ) + ('H3 SURVIVES: the amplitude is a property of the mat, not of the '
       'reference. Report it against the matched-control distribution rather '
       'than against a single reference zone.'
       if PASSED else
       'H3 FALLS: the amplitude is indistinguishable from what an arbitrary '
       'matched control produces, so it is a control-selection artefact. '
       'Remove the seasonal result from the paper and restructure around the '
       'instrumental limit and the protocol.'))

,distribution,n,median,iqr,p95
0,A - matched control,25,3.115,0.482,3.506
1,matched null (ctrl-ctrl),300,0.855,1.084,2.025


  A - matched control : median 3.12 mm  IQR 0.48
  matched null        : median 0.86 mm  p95 2.03
  controls clearing the null p95 : 100 %
  empirical p of the median      : 0.0033
  GATE: PASSED

[L1] Against 25 land-cover-matched controls the mat gives 3.12 mm (IQR 0.48); the matched null has median 0.86 mm and p95 2.03 mm. 100 % of controls clear the null, empirical p = 0.0033. H3 SURVIVES: the amplitude is a property of the mat, not of the reference. Report it against the matched-control distribution rather than against a single reference zone.



---

# L2 [SLOW] — Erode the mat  *(the untested high-value test)*

Phase K eroded the 65-pixel lake, where removing one pixel of border removes
60 % of the zone and the test cannot work. It never eroded the **mat**, which
has 499 pixels and can easily afford it.

One pass answers several objections at once, and all of them concern results
that stand whichever way the gate fell:

- **mixed-pixel contamination of the coherence deficit** — the deficit should
  *increase* on the pure interior, strengthening the most robust result;
- **the margin question** (R1 M6) by interior-versus-ring difference, instead
  of the rank test that was invalid on clustered pixels;
- **closure dispersion** on the interior, expected higher.

The radial profile already suggests which way it goes: interior plateau ≈ 0.40,
exterior peak ≈ 0.47.

In [ ]:
rows=[]
for it in (0,1,2):
    A = zones['A'] if it==0 else ref.erode_zone(zones['A'], it)
    n = int(A.values.sum())
    if n < 50:
        rows.append({'erosion_px':it,'n_px':n,'note':'too few pixels'}); continue
    z = {**zones, 'A': A}
    dfz,_ = coherence_by_zone_stream(CROPPED, pairs, z, template)
    st = paired_zone_diff(dfz, 'A', 'C')
    amp = seasonal_amplitude(invert_aggregate(
            aggregate_unwrapped(unw, corr, z, 'A', 'C')))
    cb = closure_bias_by_zone(wrapped, z, max_triplets=600,
                              zone_names=('A','C')).set_index('zone')
    rows.append({'erosion_px':it,'n_px':n,
                 'delta_mean':st.get('delta_mean'),
                 'frac_a_lower':st.get('frac_a_lower'),
                 'amplitude_mm':amp['amplitude_mm'],
                 'phase_doy':amp['phase_doy'],
                 'closure_disp_A':float(cb.loc['A','median_abs_rad']),
                 'closure_disp_C':float(cb.loc['C','median_abs_rad']),
                 'note':''})
    print(f"  erosion {it}: {n} px, delta {st.get('delta_mean'):+.4f}, "
          f"amplitude {amp['amplitude_mm']:.2f} mm, "
          f"closure A/C {cb.loc['A','median_abs_rad']:.3f}/"
          f"{cb.loc['C','median_abs_rad']:.3f}")
er = pd.DataFrame(rows); display(er.round(4))
er.to_csv(OUT/'LT04_erode_mat.csv', index=False)

  erosion 0: 499 px, delta -0.0809, amplitude 3.29 mm, closure A/C 0.683/0.212


  erosion 1: 356 px, delta -0.0832, amplitude 3.53 mm, closure A/C 0.712/0.212


  erosion 2: 233 px, delta -0.0789, amplitude 3.78 mm, closure A/C 0.761/0.212


,erosion_px,n_px,delta_mean,frac_a_lower,amplitude_mm,phase_doy,closure_disp_A,closure_disp_C,note
0,0,499,-0.0809,0.8933,3.286,104.2,0.6833,0.2124,
1,1,356,-0.0832,0.8567,3.528,106.6,0.7119,0.2124,
2,2,233,-0.0789,0.8174,3.781,108.3,0.7612,0.2124,


In [ ]:
def _l2_reading(ok):
    """Read the erosion series properly.

    Comparing only the first and last rows is too crude: eroding also shrinks
    the sample, so every estimate gets noisier. What matters is whether the
    deficit moves BEYOND that added noise, and whether closure dispersion --
    which has no competing explanation -- moves monotonically."""
    d = ok.delta_mean.values; n = ok.n_px.values
    swing = (d.max()-d.min())/abs(d.mean())
    clo = (ok.closure_disp_A/ok.closure_disp_C).values if 'closure_disp_A' in ok else None
    mono = clo is not None and bool(np.all(np.diff(clo) > 0))
    a = ok.amplitude_mm.values
    amp_obs = a[-1]/a[0]; amp_noise = np.sqrt(n[0]/n[-1])
    out = (f"Deficit varies by only {100*swing:.0f} % of its mean across the "
           f"series and is non-monotonic, so it is STABLE under erosion -- "
           f"not an artefact of mixed border pixels. ")
    if mono:
        out += (f"Closure dispersion A/C rises monotonically "
                f"{clo[0]:.2f} -> {clo[-1]:.2f}, the expected direction, "
                f"which strengthens H2 on the pure interior. ")
    out += (f"Amplitude rose x{amp_obs:.2f} where the shrinking sample alone "
            f"predicts x{amp_noise:.2f}, so it rose LESS than the noise floor.")
    return out

ok = er.dropna(subset=['delta_mean']) if 'delta_mean' in er else er.iloc[0:0]
if len(ok) >= 2:
    d0, d1 = ok.delta_mean.iloc[0], ok.delta_mean.iloc[-1]
    a0, a1 = ok.amplitude_mm.iloc[0], ok.amplitude_mm.iloc[-1]
    fig,ax=plt.subplots(1,2,figsize=(11,3.8))
    ax[0].plot(ok.erosion_px, ok.delta_mean,'o-',color=ZONE_COLORS['A'])
    ax[0].set_xlabel('pixels eroded'); ax[0].set_ylabel('paired mean coh(A) - coh(C)')
    ax[0].set_title('(a) Coherence deficit on the pure interior')
    ax[1].plot(ok.erosion_px, ok.amplitude_mm,'o-',color=ZONE_COLORS['A'])
    ax[1].set_xlabel('pixels eroded'); ax[1].set_ylabel('seasonal amplitude (mm)')
    ax[1].set_title('(b) Amplitude on the pure interior')
    save_figure(fig,'L02_erode_mat',OUT)
    verdict('L2', (
      f"Eroding the mat from {ok.n_px.iloc[0]} to {ok.n_px.iloc[-1]} px moves the "
      f"paired coherence deficit from {d0:+.4f} to {d1:+.4f} and the seasonal "
      f"amplitude from {a0:.2f} to {a1:.2f} mm. "
      + (f"Closure dispersion A/C goes from "
         f"{ok.closure_disp_A.iloc[0]/ok.closure_disp_C.iloc[0]:.2f}x to "
         f"{ok.closure_disp_A.iloc[-1]/ok.closure_disp_C.iloc[-1]:.2f}x. "
         if 'closure_disp_A' in ok else '')
      ) + _l2_reading(ok))
else:
    verdict('L2','Erosion left too few pixels to test; report as a limitation.')


[L2] Eroding the mat from 499 to 233 px moves the paired coherence deficit from -0.0809 to -0.0789 and the seasonal amplitude from 3.29 to 3.78 mm. Closure dispersion A/C goes from 3.22x to 3.58x. Deficit varies by only 5 % of its mean across the series and is non-monotonic, so it is STABLE under erosion -- not an artefact of mixed border pixels. Closure dispersion A/C rises monotonically 3.22 -> 3.58, the expected direction, which strengthens H2 on the pure interior. Amplitude rose x1.15 where the shrinking sample alone predicts x1.46, so it rose LESS than the noise floor.



---

# L3 [SLOW] — Subdivide the reference  *(makes M8 conclusive)*

Phase K subdivided the mat and found the amplitude flat, which we reported as
evidence that the signal is spatially coherent. It is not conclusive on its
own: every sub-patch of the mat keeps the **same** reference, so a cycle
carried by the *reference* would be equally flat.

Subdividing the reference at fixed target separates the two.

In [ ]:
sizes = tuple(s for s in (nC, nC//2, nC//4, 60) if s >= 40)
sub_t = ref.amplitude_vs_size(unw,corr,zones,sizes=sizes,n_draws=10,subdivide='target')
sub_r = ref.amplitude_vs_size(unw,corr,zones,sizes=sizes,n_draws=10,subdivide='reference')
both = pd.concat([sub_t, sub_r]); both.to_csv(OUT/'LT05_subdivision_both.csv',index=False)
piv = both.groupby(['subdivided','n_px']).amplitude_mm.agg(['median','std'])
display(piv.round(3))

fig,ax=plt.subplots(figsize=(6.6,4))
for name,g,c in [('target (mat)',sub_t,ZONE_COLORS['A']),
                 ('reference (grassland)',sub_r,ZONE_COLORS['C'])]:
    m=g.groupby('n_px').amplitude_mm.median()
    ax.plot(m.index,m.values,'o-',color=c,label=f'subdividing {name}')
ax.set_xscale('log'); ax.set_xlabel('pixels aggregated')
ax.set_ylabel('seasonal amplitude (mm)'); ax.legend(fontsize=8)
ax.set_title('L3 — which side carries the cycle?')
save_figure(fig,'L03_subdivide_both',OUT)

def drift(g):
    m=g.groupby('n_px').amplitude_mm.median()
    return float(m.iloc[-1]/m.iloc[0]) if len(m)>1 and m.iloc[0] else float('nan')
dt, dr = drift(sub_t), drift(sub_r)
verdict('L3', (
  f"Amplitude ratio smallest/largest patch: {dt:.2f} subdividing the mat, "
  f"{dr:.2f} subdividing the reference. "
  ) + ('Flat under BOTH -> both terms are spatially coherent and M8 is '
       'conclusive.' if (abs(dt-1)<0.5 and abs(dr-1)<0.5) else
       'Not flat under both -> the subdivision evidence is ambiguous about '
       'which zone carries the cycle, and M8 must be reported as such.'))

median    std
subdivided n_px               
A          60     3.648  0.703
           99     3.724  0.505
           199    3.624  0.269
           398    3.406  0.087
C          60     3.223  0.717
           99     3.156  0.501
           199    3.234  0.210
           398    3.286  0.000


[L3] Amplitude ratio smallest/largest patch: 0.93 subdividing the mat, 1.02 subdividing the reference. Flat under BOTH -> both terms are spatially coherent and M8 is conclusive.



---

# L4 — Closed triplets  *(R1 S7, one hour, closes an "unresolved")*

The number of closed triangles is fully determined by the pair list. Counting
it two independent ways — trace(A³)/6 and direct enumeration — removes an open
item that looked far worse than its actual difficulty.

In [ ]:
tri = ref.count_closed_triplets(pairs)
for k,v in tri.items(): print(f'  {k:26s} {v}')
pd.DataFrame([tri]).to_csv(OUT/'LT06_triplets.csv',index=False)
verdict('L4', (
  f"{tri['triplets_trace']} closed triplets from {tri['n_pairs']} pairs over "
  f"{tri['n_dates']} dates; trace(A^3)/6 and enumeration agree "
  f"({tri['agree']}). A random graph of the same density would close "
  f"{tri['random_graph_expectation']:.0f}. The network closes "
  f"{tri['ratio_to_random']:.1f}x MORE triangles than random, not fewer: "
  f"constrained HyP3 baselines cluster pairs in time, and temporally "
  f"clustered pairs close more triplets. The premise that 518 is low for "
  f"this network does not hold, and the count is fully reproducible."))

  n_dates                    90
  n_pairs                    356
  triplets_trace             518
  triplets_enumerated        518
  agree                      True
  random_graph_expectation   82.50995884773666
  ratio_to_random            6.278030037027577

[L4] 518 closed triplets from 356 pairs over 90 dates; trace(A^3)/6 and enumeration agree (True). A random graph of the same density would close 83. The network closes 6.3x MORE triangles than random, not fewer: constrained HyP3 baselines cluster pairs in time, and temporally clustered pairs close more triplets. The premise that 518 is low for this network does not hold, and the count is fully reproducible.



---

# L5 — Amplitude without the network inversion  *(closes the M4 hole)*

We conceded that immunity to unwrapping errors does not survive the inversion
to date-referenced values — and the published amplitude comes from exactly
that inversion. Fitting the annual cycle directly on per-pair differences
avoids it. Safe here because 3.3 mm is ≈ 0.75 rad two-way, well inside ±π.

In [ ]:
dd = aggregate_unwrapped(unw, corr, zones, 'A', 'C')
wfit = ref.wrapped_seasonal_amplitude(dd)
for k,v in wfit.items():
    print(f'  {k:18s} {v:.4f}' if isinstance(v,float) else f'  {k:18s} {v}')
pd.DataFrame([wfit]).to_csv(OUT/'LT07_wrapped_amplitude.csv',index=False)
# The wrapped fit is a far noisier estimator, so the two must be compared
# against ITS uncertainty, not against a flat percentage. SE on each harmonic
# coefficient ~ rms / sqrt(n/2).
se_w = wfit['rms_residual_mm']/np.sqrt(wfit['n_pairs']/2)
sig_w = abs(wfit['amplitude_mm']-obs['amplitude_mm'])/se_w
rel = abs(wfit['amplitude_mm']-obs['amplitude_mm'])/obs['amplitude_mm']
print(f'  SE per harmonic coefficient : {se_w:.2f} mm')
print(f'  difference                  : {sig_w:.2f} sigma of this estimator')
verdict('L5', (
  f"Fitted without the network inversion: {wfit['amplitude_mm']:.2f} mm "
  f"against {obs['amplitude_mm']:.2f} mm with it ({100*rel:.0f} % difference, "
  f"RMS residual {wfit['rms_residual_mm']:.2f} mm). "
  f"SE {se_w:.2f} mm, so they differ by {sig_w:.1f} sigma of the wrapped "
  f"estimator, whose r2 is {wfit['r2']:.4f}. "
  ) + ('The two are CONSISTENT within the wrapped estimator uncertainty. '
       'It is too noisy to confirm the inversion-based value independently, '
       'but it does not contradict it -- the test is inconclusive, not '
       'negative.' if sig_w < 2 else
       'They differ by more than 2 sigma even of this noisy estimator, so '
       'the inversion is doing real work and the amplitude is NOT covered '
       'by the wrapped-phase immunity argument.'))

  amplitude_mm       2.3202
  phase_doy          105.5856
  trend_mm_yr        -2.0171
  n_pairs            356
  r2                 0.0053
  rms_residual_mm    6.6273
  SE per harmonic coefficient : 0.50 mm
  difference                  : 1.94 sigma of this estimator

[L5] Fitted without the network inversion: 2.32 mm against 3.29 mm with it (29 % difference, RMS residual 6.63 mm). SE 0.50 mm, so they differ by 1.9 sigma of the wrapped estimator, whose r2 is 0.0053. The two are CONSISTENT within the wrapped estimator uncertainty. It is too noisy to confirm the inversion-based value independently, but it does not contradict it -- the test is inconclusive, not negative.



---

# L6 [VERY SLOW] — Corrected null at 5 000 draws  *(only if the gate passed)*

The *p* that matters now comes from the reference-corrected null, and it rests
on 184 draws with 7 exceedances — exact binomial 95 % interval **[0.015,
0.077]**, which straddles 0.05. A conclusion cannot rest on a *p* whose own
interval contains the threshold.

This is machine time and nothing else. Skip it if L1 failed.

In [ ]:
if not PASSED:
    print('Gate failed -- skipping L6; the p-value of a withdrawn result is moot.')
else:
    big = ref.null_against_real_reference(unw,corr,zones,template,nA,
                                          reference='C',n_trials=5000)
    pb = empirical_pvalue(obs['amplitude_mm'], big.amplitude_mm)
    k = int((big.amplitude_mm >= obs['amplitude_mm']).sum())
    from scipy import stats as _st
    lo = _st.beta.ppf(.025, k, len(big)-k+1) if k else 0.0
    hi = _st.beta.ppf(.975, k+1, len(big)-k)
    big.to_csv(OUT/'LT08_null_5000.csv', index=False)
    print(f'{len(big)} draws, {k} exceedances -> p = {pb["p_value"]:.4f}')
    verdict('L6', f"With {len(big)} draws of the reference-corrected null, "
            f"p = {pb['p_value']:.4f}, exact 95 % CI [{lo:.4f}, {hi:.4f}]. "
            + ('The interval no longer straddles 0.05.' if hi < 0.05 else
               'The interval STILL straddles 0.05 -- the detection is not '
               'secure at the conventional threshold and must be reported so.'))

4614 draws, 119 exceedances -> p = 0.0260

[L6] With 4614 draws of the reference-corrected null, p = 0.0260, exact 95 % CI [0.0214, 0.0308]. The interval no longer straddles 0.05.



---

# L7 — The 27 pixels, tested properly  *(R1 M6)*

R1 M6 objected that our Mann–Whitney treated 27 *clustered* pixels as 27
independent observations at a 160 m correlation length. The premise turns out
to be wrong, and it is worth saying so plainly: the 27 pixels form **15
separate fragments** — mostly isolated pixels and pairs — spread across a
1000 × 1480 m box. They are not a cluster. Most fragments are separated by more
than the correlation length, so the effective sample size is nearer 15 than 1,
and the Mann–Whitney was inflated by perhaps a factor of two rather than by a
factor of 27.

It was still not a valid test — the pixels were selected by a threshold on
temporal coherence and scored against a spatially autocorrelated distance
field — so a permutation test remains the right instrument. But the correction
is smaller than we conceded, and conceding too much is its own kind of error.

The null is built in three stages, strongest first: a rigid toroidal shift
(exact shape); failing that, a **component-matched** draw reproducing the
fragment count and fragment sizes at random positions; failing that, compact
blobs of equal pixel count. The mode actually used is reported, because they
are not equally strong.

Note none of this can separate the two physical explanations — an anchored edge
that holds phase, and border pixels contaminated by stable surrounding terrain.
The median distance is one pixel, which is what both predict.

**The tail is one-sided, and deliberately so.** The hypothesis is directional:
the surviving pixels sit *closer to the margin* than chance. Signed distance is
negative inside the zone, so that means a *higher* value. The null of positions
is strongly left-skewed — most of the zone lies far from its own boundary — so a
two-sided test on |deviation from the null median| counts the long inward tail
against us and can report a null result for a cluster that is nonetheless more
marginal than 95 % of null draws. Both tails are printed below so the choice is
visible rather than buried.

**Watch the null mode.** Fairness is judged on *intrinsic* shape only: the
number of fragments and how elongated each fragment is, since those fix what
the set can physically reach. The spread of the set as a whole is deliberately
**not** matched — for a fragmented set that is a property of position, and
position is the hypothesis, so a null reproducing it would test nothing. If the
null cannot match the intrinsic shape, the descriptors below say so and the
verdict downgrades itself automatically.

In [ ]:
tcoh=None
try:
    _t=xr.open_dataset(DRIVE/'phaseE2_evd.nc')['temporal_coherence']
    tcoh=to_grid(_t,template) if _t.shape!=template.shape else _t
except Exception as e: print('EVD result unavailable:',e)

if tcoh is not None:
    good = (tcoh.values>=0.7) & zones['A'].values
    subset = xr.DataArray(good, coords=template.coords, dims=template.dims)
    r = ref.toroidal_permutation_test(subset, zones['A'], sd, n_trials=4000,
                                      tail='greater')
    for k,v in r.items():
        if k!='nulls': print(f'  {k:18s} {v}')
    if r.get('shape_warning'): print('\n  ***', r['shape_warning'])
    pd.DataFrame([{k:v for k,v in r.items()
                   if k not in ('nulls','shape_observed','shape_null_median')}]).to_csv(
        OUT/'LT09_marginal_permutation.csv',index=False)
    if r.get('n_null',0):
        fig,ax=plt.subplots(figsize=(6,3.6))
        ax.hist(r['nulls'],bins=50,color='gray',alpha=.8,label='toroidal null')
        ax.axvline(r['observed'],c='tab:red',lw=2,label=f"observed {r['observed']:.0f} m")
        ax.set_xlabel('median signed distance to boundary (m)')
        ax.legend(fontsize=8); ax.set_title('L7 — margin concentration, autocorrelation-preserving')
        save_figure(fig,'L04_marginal_permutation',OUT)
    if not r.get('n_null'):
        verdict('L7', f"NO RESULT: {r['n_subset']} pixels, but no null could "
                f"be constructed ({r.get('note','')}). This is NOT evidence "
                f"against the marginal concentration -- the test did not run. "
                f"Report the observation without a p-value.")
    else:
        verdict('L7', (
          f"{r['n_subset']} surviving pixels: median distance {r['observed']:.0f} m "
          f"against a null of {r['null_median']:.0f} m "
          f"[{r['null_p05']:.0f}, {r['null_p95']:.0f}]. One-sided "
          + (f"p < {r['p_floor']:.5f} (0 of {r['n_null']} draws reached it, "
             f"so the permutation bounds p from above and cannot estimate it)"
             if r.get('p_is_censored_at_floor') else
             f"p = {r['p_greater']:.4f} ({r['n_null']} draws)")
          + f", two-sided {r['p_two_sided']:.4f}, reported for transparency; "
          + f"the null is left-skewed so the two-sided statistic is not the "
          + f"one the directional hypothesis asks for. Null mode: {r['mode']}. "
          ) + (('SHAPE-CONFOUNDED. ' + r['shape_warning'] + ' The direction is '
                'right and the Mann-Whitney was certainly invalid, but this test '
                'does not cleanly establish the position effect either. Report '
                'the concentration as an observation, not as a significant '
                'permutation result.') if r.get('shape_warning') else
               'Marginal concentration SURVIVES an autocorrelation-preserving '
               'test -- but a mixed-pixel explanation remains unexcluded at this '
               'resolution, so it stays an observation with two live readings.'
               if r['p_greater'] < 0.05 else
               'The concentration does NOT survive once clustering is respected; '
               'the Mann-Whitney p of 0.004 was inflated by treating neighbours '
               'as independent.'))

  n_subset           27
  observed           -40.0
  mode               component-matched (count, fragment sizes and compactness preserved; position randomised)
  tail               greater
  n_rigid_accepted   0
  p_is_censored_at_floor False
  shape_observed     {'n_px': 27, 'n_components': 15, 'radius_of_gyration': 11.813595664487824, 'component_radius_of_gyration': 0.33628953425123514, 'bbox_h': 25, 'bbox_w': 37, 'fill_fraction': 0.02918918918918919}
  shape_null_median  {'n_components': 13.0, 'radius_of_gyration': 10.624660084936316, 'component_radius_of_gyration': 0.3764943993076858, 'fill_fraction': 0.0406015037593985, 'component_rg_ratio_observed_over_null': 1.0, 'rg_ratio_observed_over_null': 1.1119033992661267}
  shape_warning      
  zone_statistic     -113.13708498984761
  n_null             4000
  null_median        -89.44271909999159
  null_p05           -144.22205101855957
  null_p95           -56.568542494923804
  p_value            0.000999750062484379
  p_greater     


[L7] 27 surviving pixels: median distance -40 m against a null of -89 m [-144, -57]. One-sided p = 0.0010 (4000 draws), two-sided 0.0992, reported for transparency; the null is left-skewed so the two-sided statistic is not the one the directional hypothesis asks for. Null mode: component-matched (count, fragment sizes and compactness preserved; position randomised). Marginal concentration SURVIVES an autocorrelation-preserving test -- but a mixed-pixel explanation remains unexcluded at this resolution, so it stays an observation with two live readings.



---

# L8 — Three intervals, not one  *(R1 S4, refined)*

An i.i.d. bootstrap over dates destroys the temporal sampling structure, and we
are fitting an annual cycle whose conditioning depends on how evenly the year
is covered. Resamples that over- or under-represent parts of the year degrade
the fit and **inflate** the interval, so [0.58, 7.32] is probably conservative.

Reporting the analytic interval alongside it shows how much.

In [ ]:
series = invert_aggregate(dd)
t = (pd.to_datetime(series['date']) - pd.to_datetime(series['date']).min()).dt.days/365.25
X = np.column_stack([np.ones(len(t)), t, np.cos(2*np.pi*t), np.sin(2*np.pi*t)])
y = series['disp_mm'].values
beta, *_ = np.linalg.lstsq(X, y, rcond=None)
resid = y - X@beta; dof = max(len(y)-X.shape[1], 1)
cov = np.var(resid, ddof=X.shape[1]) * np.linalg.inv(X.T@X)
a,b = beta[2], beta[3]; amp = np.hypot(a,b)
# delta method on sqrt(a^2+b^2)
g = np.array([a,b])/amp
se_amp = float(np.sqrt(g @ cov[2:,2:] @ g))
boot = ref.bootstrap_amplitude_ci(dd, n_boot=2000, seed=0)
cmp_ = pd.DataFrame([
  {'method':'analytic (harmonic fit covariance)','amplitude_mm':amp,
   'ci_low':amp-1.96*se_amp,'ci_high':amp+1.96*se_amp},
  {'method':'i.i.d. date bootstrap (conservative)','amplitude_mm':boot['median'],
   'ci_low':boot['ci_low'],'ci_high':boot['ci_high']}])
display(cmp_.round(3)); cmp_.to_csv(OUT/'LT10_intervals.csv',index=False)
w_an = cmp_.ci_high[0]-cmp_.ci_low[0]; w_bs = cmp_.ci_high[1]-cmp_.ci_low[1]
verdict('L8', (
  f"Analytic 95 % CI [{cmp_.ci_low[0]:.2f}, {cmp_.ci_high[0]:.2f}] mm "
  f"(width {w_an:.2f}) against the date bootstrap "
  f"[{cmp_.ci_low[1]:.2f}, {cmp_.ci_high[1]:.2f}] (width {w_bs:.2f}). "
  ) + (f'The bootstrap is {w_bs/w_an:.1f}x wider, consistent with the '
       'i.i.d. scheme degrading the annual-cycle conditioning; report both.'
       if w_bs > 1.3*w_an else
       'The two agree, so the bootstrap interval is not an artefact of the '
       'resampling scheme and should be reported as the interval.'))

,method,amplitude_mm,ci_low,ci_high
0,analytic (harmonic fit covariance),3.286,2.224,4.349
1,i.i.d. date bootstrap (conservative),3.310,0.585,7.322



[L8] Analytic 95 % CI [2.22, 4.35] mm (width 2.13) against the date bootstrap [0.58, 7.32] (width 6.74). The bootstrap is 3.2x wider, consistent with the i.i.d. scheme degrading the annual-cycle conditioning; report both.



---

# Summary

In [ ]:
import textwrap
print('='*78)
print(f"  GATE (L1): {'PASSED -- H3 survives' if PASSED else 'FAILED -- H3 falls'}")
print('='*78)
for tag,txt in VERDICTS:
    print(f'\n[{tag}]')
    print(textwrap.fill(txt,76,initial_indent='  ',subsequent_indent='  '))
print('\n'+'='*78)
pd.DataFrame(VERDICTS,columns=['test','verdict']).to_csv(
    OUT/'phaseL_verdicts.csv',index=False)
print(f'{len(VERDICTS)} verdicts -> {OUT}/phaseL_verdicts.csv')

  GATE (L1): PASSED -- H3 survives

[L1]
  Against 25 land-cover-matched controls the mat gives 3.12 mm (IQR 0.48);
  the matched null has median 0.86 mm and p95 2.03 mm. 100 % of controls
  clear the null, empirical p = 0.0033. H3 SURVIVES: the amplitude is a
  property of the mat, not of the reference. Report it against the matched-
  control distribution rather than against a single reference zone.

[L2]
  Eroding the mat from 499 to 233 px moves the paired coherence deficit from
  -0.0809 to -0.0789 and the seasonal amplitude from 3.29 to 3.78 mm.
  Closure dispersion A/C goes from 3.22x to 3.58x. Deficit varies by only 5
  % of its mean across the series and is non-monotonic, so it is STABLE
  under erosion -- not an artefact of mixed border pixels. Closure
  dispersion A/C rises monotonically 3.22 -> 3.58, the expected direction,
  which strengthens H2 on the pure interior. Amplitude rose x1.15 where the
  shrinking sample alone predicts x1.46, so it rose LESS than the noise
  fl

In [ ]:
!git add -A docs/paper
!git commit -m 'phaseL: gate result and supporting tests' || echo 'nothing to commit'
!git push origin {BRANCH} || true

Author identity unknown

*** Please tell me who you are.

Run

  git config --global user.email "you@example.com"
  git config --global user.name "Your Name"

to set your account's default identity.
Omit --global to set the identity only in this repository.

fatal: unable to auto-detect email address (got 'root@175bce37f434.(none)')
nothing to commit


fatal: could not read Username for 'https://github.com': No such device or address


---

### After the gate

**If L1 passed** — H3 stays, reported against the matched-control distribution
rather than a single reference. Finish L6 and L8, then the remaining items are
the literature calibration and the descending-track reproduction.

**If L1 failed** — the seasonal result comes out. What remains is a stronger,
simpler paper than the one you have: the coherence deficit at matched cover,
the failure of every available per-pixel chain, the ×3.2 closure dispersion,
and a weak-signal protocol demonstrated by the fact that it withdrew its own
authors' conclusions. L2, L3, L4 and L7 all still apply, and none of it needs
field data — which also removes the water-table and coring dependencies from
the critical path.

Either way, two items remain outside this notebook:

- **the moisture-induced-phase literature**, which decides whether the closure
  result is underpowered or discriminating and negative — not deferrable,
  because §4.8 cannot be written without it;
- **the EGMS sign check**, an afternoon, without which H4's first inferential
  link is unverified by our own admission.

In [ ]:
# --- Archive this execution ---------------------------------------------
# <Drive>/runs/phaseL/<timestamp>_<git sha>/ : commit, environment, parameters,
# products. Never overwrites a previous run.
run = ctx.archive(params={}, products={})   # name this phase's outputs here
print("archived:", run)


INFO    archived run: /content/drive/MyDrive/insar_rzecin/runs/phaseL/20260906T213344Z_db85693


archived: /content/drive/MyDrive/insar_rzecin/runs/phaseL/20260906T213344Z_db85693
